resources/type-system/normalization.md

```diff
@@ -0,0 +1,93 @@
+# Dart 2.0 Type Normalization
+
+leafp@google.com
+
+With union, intersection, and bottom types, there are types which are
+syntactically different, but are equal in the sense that they are mutual
+subtypes.  This document defines a proposed normalization procedure for choosing
+a canonical representative of equivalence classes of types.  Such a procedure
+might provide a basis for choosing between mutual subtypes when computing upper
+and lower bounds.  It might also provide a more efficient implementation of type
+equality, since normal forms for types could be eagerly or lazy computed and
+cached.
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+We assume that type aliases are fully expanded, and that prefixed types are
+resolved to a canonical name.
+
+## Normalization
+
+The **NORM** relation defines the canonical representative of classes of
+equivalent types.  In the absence of legacy (*) types, it should be the case
+that for any two types which are mutual subtypes, their normal forms are either
+one of the primitive top types, or are syntactically identical.
+
+This is based on the following equations:
+- `T?? == T?`
+- `T?* == T?`
+- `T*? == T?`
+- `T** == T*`
+- `Null? == Null`
```

Comment:
That should probably be true.
A `*` type is different from its base type and its nullable base type because it can be used as either, co- and contra-variantly.
However if `Null?` and `Null` are equivalent, then that distinction doesn't matter and `Null*` should be equivalent to both.